### PR TITLE
Update MySQL JDBC driver to 8.0.29

### LINKED ISSUE #15332: The mysql system.query table function cannot obtain ResultSetMetadata for queries having WITH statements
Queries such as 
```
SELECT * FROM TABLE(mysql.system.query(
query => 'with customers_who_ordered as (select distinct custkey from tpch.orders)  
select custkey, name from tpch.customer 
where custkey not in (select * from customers_who_ordered)'));
```
Will fail with
```
Query 20221206_012713_00014_crkwa failed: com.google.common.util.concurrent.UncheckedExecutionException: 
java.lang.UnsupportedOperationException: Query not supported: ResultSetMetaData not available for query: 
with customers_who_ordered as (select distinct custkey from tpch.orders)  
select custkey, name from tpch.customer where custkey not in (select * from customers_who_ordered)
```
This appears to be a limitation of the 8.0.22 JDBC driver currently used by the connector. CC @ebyhr

### PR DESCRIPTION
**Pull request by hashhar**
## Description

Update MySQL JDBC driver to 8.0.29

Fixes #15332 Closes #12692

## Additional context and related issues

Update MySQL JDBC driver to 8.0.29

## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
# MySQL
* Update MySQL JDBC driver to 8.0.29. This fixes failure when `query` table function contains `WITH` clause. ({issue}`15332`)
```

### CODE REVIEW

**Comment by hashhar** [comment]
Alternative to https://github.com/trinodb/trino/pull/12692, let's see what tests fail.

**Comment by hashhar** [comment]
Same failures as in https://github.com/trinodb/trino/pull/14917, negative date related tests.

I guess we'll need to add validation on Trino side to disallow negative values until the driver is fixed. I'll see how we can report a bug to the driver - I just left a comment at https://github.com/mysql/mysql-connector-j/commit/60553fd2557267f3184b77e1832025f80d79e89a#r88985720 in past.

**Comment by hashhar** [comment]
Got it working, added a test to verify driver bug too so that once we update to working driver we can get notified.

I've started the process to file the bug to MySQL - waiting to hear back.

File: plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java [hunk 1]
```diff
@@ -456,6 +471,62 @@ public Optional<ColumnMapping> toColumnMapping(ConnectorSession session, Connect
         return Optional.empty();
     }
 
+    private LongWriteFunction mySqlDateWriteFunctionUsingLocalDate()
+    {
+        LocalDate minMySqlDate = LocalDate.of(0, 1, 1);
+        return LongWriteFunction.of(Types.DATE, (statement, index, epochDay) -> {
+            LocalDate localDate = LocalDate.ofEpochDay(epochDay);
+            if (localDate.isBefore(minMySqlDate)) {
+                throw new TrinoException(INVALID_ARGUMENTS, format("Value must be after %s in MySQL: '%s'", minMySqlDate, localDate));
```
Comment (ebyhr) [inline plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java:480]: Does "must be after %s" mean exclusive? How about "on or after" or "after or equal to"? 
Please ignore this comment if it doesn't make sense.

File: plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java [hunk 2]
```diff
@@ -456,6 +471,62 @@ public Optional<ColumnMapping> toColumnMapping(ConnectorSession session, Connect
...
+            }
+            statement.setObject(index, localDate);
```
Comment (ebyhr) [inline plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java:482]: If we use `PreparedStatement.setString` with bind expression, there's no need to validate in connector side?

File: plugin/trino-mysql/src/test/java/io/trino/plugin/mysql/TestMySqlLegacyConnectorTest.java [hunk 3]
```diff
@@ -167,4 +167,24 @@ protected void verifyColumnNameLengthFailurePermissible(Throwable e)
...
+        assertThatThrownBy(super::testNativeQueryWithClause)
+                .hasStackTraceContaining("Failed to get table handle for prepared query. You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version");
+    }
+
+    @Test
+    @Override
+    public void testDateYearOfEraPredicate()
+    {
+        // MySQL 5 returns incorrect results instead of an empty result with drivers newer than 8.0.23
+        // This is overriden since the connector validates the dates now and throws on such case
```
Comment (ebyhr) [inline plugin/trino-mysql/src/test/java/io/trino/plugin/mysql/TestMySqlLegacyConnectorTest.java:184]: nit: Fix typo overriden → overridden

**Review by ebyhr — APPROVED** [review]

File: plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java (hunk 2 above)
Comment (hashhar) [reply plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java:482]: Good idea, let me try that.

**Comment by hashhar** [comment]
@ebyhr Addressed comments as fixups.

It's using `bindExpression` now - much cleaner - thanks for the suggestion. The error message changed slightly since now the server sees failure in `CAST` instead of inserting to column.

**Comment by hashhar** [comment]
This is ready to review. PTAL @ebyhr @electrum

**Review by ebyhr — APPROVED** [review]

**Pull request merged**